sklearn/experimental/_column_transformer.py

```diff
@@ -0,0 +1,212 @@
+
+
+class ColumnTransformer(FeatureUnion):
+    """Applies transformers to columns of an array or pandas DataFrame.
+
+    This estimator applies transformer objects to columns or fields of the
+    input, then concatenates the results. This is useful for heterogeneous or
+    columnar data, to combine several feature extraction mechanisms into a
+    single transformer.
+
+    Read more in the :ref:`User Guide <column_transformer>`.
+
+    Parameters
+    ----------
+    transformer_list : list of tuples
+        List of (name, transformer, column) tuples specifying the transformer
+        objects to be applied to subsets of the data. The columns can be
+        specified as a scalar or slice/list (for multiple columns) of integer
+        or string values. Integers are interpreted as the positional columns,
+        strings as the keys (column labels) of `X`.
+        When passing a single column to a transformer that expects 2D input
+        data, the column should be specified a list of one element.
+
+    n_jobs : int, optional
+        Number of jobs to run in parallel (default 1).
+
+    transformer_weights : dict, optional
+        Multiplicative weights for features per transformer.
+        Keys are transformer names, values the weights.
+
+    Examples
+    --------
+    >>> from sklearn.experimental import ColumnTransformer
+    >>> from sklearn.preprocessing import Normalizer
+    >>> union = ColumnTransformer(
+    ...     [("norm1", Normalizer(norm='l1'), [0, 1]),
+    ...      ("norm2", Normalizer(norm='l1'), [2, 3])])
+    >>> X = np.array([[0., 1., 2., 2.],
+    ...               [1., 1., 0., 1.]])
+    >>> union.fit_transform(X)    # doctest: +NORMALIZE_WHITESPACE
+    array([[ 0. ,  1. ,  0.5,  0.5],
+           [ 0.5,  0.5,  0. ,  1. ]])
+
+    """
+    def _iter(self, X=None, skip_none=True):
+        """Generate (name, trans, column, weight) tuples
+        """
+        get_weight = (self.transformer_weights or {}).get
+        return ((name, trans, _get_column(X, column), get_weight(name))
+                for name, trans, column in self.transformer_list
+                if not skip_none or trans is not None)
+
+    def _update_transformer_list(self, transformers):
+        transformers = iter(transformers)
+        self.transformer_list[:] = [
+            (name, None if old is None else next(transformers), column)
+            for name, old, column in self.transformer_list
+        ]
+
+    def fit(self, X, y=None):
+        """Fit all transformers using X.
+
+        Parameters
+        ----------
+        X : array-like or DataFrame of shape [n_samples, n_features]
+            Input data, of which specified subsets are used to fit the
+            transformers.
+
+        y : array-like, shape (n_samples, ...), optional
+            Targets for supervised learning.
+
+        Returns
+        -------
+        self : ColumnTransformer
+            This estimator
+        """
```

Comment:
Current `safe_indexing` (for first axis) also has no special handling for sparse arrays. It works for those sparse arrays that support indexing, and raises an error for those who don't (the same error as if you would just try to index it)